x-pack/plugins/observability_ai_assistant/server/functions/context.ts

```diff
@@ -77,49 +63,94 @@ export function registerRecallFunction({
 
-      if (!systemMessage) {
-        throw new Error('No system message found');
-      }
+      async function getContext() {
+        const systemMessage = messages.find(
+          (message) => message.message.role === MessageRole.System
+        );
 
-      const userMessage = last(
-        messages.filter((message) => message.message.role === MessageRole.User)
-      );
+        const screenDescription = appContexts.map((context) => context.description).join('\n\n');
 
-      const nonEmptyQueries = compact(queries);
+        const content = { screen_description: screenDescription, learnings: [] };
 
-      const queriesOrUserPrompt = nonEmptyQueries.length
-        ? nonEmptyQueries
-        : compact([userMessage?.message.content]);
+        if (!isKnowledgeBaseAvailable) {
+          return { content };
+        }
 
-      const suggestions = await retrieveSuggestions({
-        userMessage,
-        client,
-        categories,
-        queries: queriesOrUserPrompt,
-      });
+        if (!systemMessage) {
+          throw new Error('No system message found');
+        }
+
+        const userMessage = last(
+          messages.filter((message) => message.message.role === MessageRole.User)
+        );
+
+        const nonEmptyQueries = compact(queries);
+
+        const queriesOrUserPrompt = nonEmptyQueries.length
+          ? nonEmptyQueries
+          : compact([userMessage?.message.content]);
+
+        queriesOrUserPrompt.push(screenDescription);
+
+        const suggestions = await retrieveSuggestions({
+          userMessage,
+          client,
+          categories,
+          queries: queriesOrUserPrompt,
+        });
+
+        if (suggestions.length === 0) {
+          return {
+            content,
+          };
+        }
+
+        const relevantDocuments = await scoreSuggestions({
+          suggestions,
+          queries: queriesOrUserPrompt,
+          messages,
+          client,
+          connectorId,
+          signal,
+          logger: resources.logger,
+        });
 
-      if (suggestions.length === 0) {
         return {
-          content: [] as unknown as Serializable,
+          content: { ...content, learnings: relevantDocuments as unknown as Serializable },
         };
       }
 
-      const relevantDocuments = await scoreSuggestions({
-        suggestions,
-        queries: queriesOrUserPrompt,
-        messages,
-        client,
-        connectorId,
-        signal,
-        logger: resources.logger,
+      return new Observable<MessageAddEvent>((subscriber) => {
+        getContext()
+          .then(({ content }) => {
+            // any data that falls within the token limit, send it automatically
+
+            const dataWithinTokenLimit = compact(
+              appContexts.flatMap((context) => context.data)
+            ).filter(
+              (data) =>
+                encode(JSON.stringify(data.value)).length <= MAX_TOKEN_COUNT_FOR_DATA_ON_SCREEN
+            );
```

Comment:
Why is this not part of the `getContext` function?

Reply:
It wasn't part of it because I originally implemented it as separate messages (`get_data_on_screen` function calls). But I was worried that that would cause the LLM to no longer call `get_data_on_screen`, so I removed it again. I've moved it to the `getContext` function.